src/main.rs

```diff
@@ -4,297 +4,165 @@ use clap::crate_authors;
 use std::io;
 use std::time::SystemTime;
 
-use clap::{App, AppSettings, Arg, Shell, SubCommand};
+use clap::{AppSettings, IntoApp, Parser, Subcommand};
+use clap_complete::{generate, Shell as CompletionShell};
 use rand::distributions::Alphanumeric;
 use rand::Rng;
+use starship::context::{Properties, Target};
 use starship::module::ALL_MODULES;
 use starship::*;
 
+fn long_version() -> &'static str {
+    let ver = Box::new(crate::shadow::clap_version());
+    Box::leak(ver).as_str()
+}
+
+#[derive(Parser, Debug)]
+#[clap(
+    author=crate_authors!(),
+    version=shadow::PKG_VERSION,
+    long_version=long_version(),
+    about="The cross-shell prompt for astronauts. ☄🌌️"
+)]
+#[clap(setting(AppSettings::SubcommandRequiredElseHelp))]
+struct Cli {
+    #[clap(subcommand)]
+    command: Commands,
+}
+
+#[derive(Subcommand, Debug)]
+enum Commands {
+    /// Create a pre-populated GitHub issue with information about your configuration
+    BugReport,
+    /// Generate starship shell completions for your shell to stdout
+    Completions {
+        #[clap(arg_enum)]
+        shell: CompletionShell,
+    },
+    /// Edit the starship configuration
+    Config {
+        /// Configuration key to edit
+        #[clap(requires = "value")]
+        name: Option<String>,
+        /// Value to place into that key
+        value: Option<String>,
+    },
+    /// Explains the currently showing modules
+    Explain(Properties),
+    ///  Prints the shell function used to execute starship
+    Init {
+        shell: String,
```

Comment:
We could instead create an enum with all shell types and derive `ArgEnum` for it

Reply:
I didn't do that because passing the shell path is also supported for this argument. Also, how to handle unknown shell names will require some thought.